wrapper/src/main/java/software/amazon/jdbc/plugin/cache/CacheMonitor.java

```diff
@@ -0,0 +1,917 @@
+    volatile StatefulRedisConnection<byte[], byte[]> roPingConnection;
+
+    // Timing tracking for ping intervals
+    volatile long rwLastPingTimestamp;
+    volatile long roLastPingTimestamp;
+
+    // Locks for ping connection management
+    final ReentrantLock rwPingLock;
+    final ReentrantLock roPingLock;
+
+    ClusterHealthState(String rwEndpoint, String roEndpoint, PingConnectionFactory connectionFactory) {
+      this.rwEndpoint = rwEndpoint;
+      this.roEndpoint = roEndpoint;
+      this.connectionFactory = connectionFactory;
+
+      // Initialize health states
+      this.rwHealthState = HealthState.HEALTHY;
+      this.roHealthState = HealthState.HEALTHY;
+
+      // Initialize counters
+      this.consecutiveRwSuccesses = new AtomicInteger(0);
+      this.consecutiveRwFailures = new AtomicInteger(0);
+      this.consecutiveRoSuccesses = new AtomicInteger(0);
+      this.consecutiveRoFailures = new AtomicInteger(0);
+      this.inFlightWriteSizeBytes = new AtomicLong(0);
+
+      // Initialize timestamps
+      this.rwLastPingTimestamp = 0;
+      this.roLastPingTimestamp = 0;
+
+      // Initialize locks
+      this.rwPingLock = new ReentrantLock();
+      this.roPingLock = new ReentrantLock();
+    }
+
+    /**
+     * Generate a unique cluster key from RW and RO endpoints.
+     */
+    String getClusterKey() {
+      if (roEndpoint == null || roEndpoint.equals(rwEndpoint)) {
+        return rwEndpoint;
+      }
+      return rwEndpoint + "|" + roEndpoint;
+    }
+
+    /**
+     * Reset consecutive counters for a specific endpoint.
+     */
+    void resetConsecutiveCounters(boolean isRw) {
+      if (isRw) {
+        consecutiveRwSuccesses.set(0);
+        consecutiveRwFailures.set(0);
+      } else {
+        consecutiveRoSuccesses.set(0);
+        consecutiveRoFailures.set(0);
+      }
+    }
+
+    /**
+     * Transition health state for a specific endpoint.
+     */
+    void transitionToState(HealthState newState, boolean isRw, String triggerReason,
+                           TelemetryCounter counter) {
+      String endpoint = isRw ? rwEndpoint : roEndpoint;
+      HealthState oldState = isRw ? rwHealthState : roHealthState;
+
+      // Update state
+      if (isRw) {
+        rwHealthState = newState;
+      } else {
+        roHealthState = newState;
+      }
+
+      resetConsecutiveCounters(isRw);
+
+      // Emit telemetry
+      if (counter != null) {
+        counter.inc();
+      }
+
+      // Log state transition
+      if (!triggerReason.startsWith("recoverable_error_")) {
+        String logMessage = String.format(
+            "[%s to %s] Cache endpoint %s (%s) health state transitioned (trigger: %s)",
+            oldState, newState, endpoint, isRw ? "RW" : "RO", triggerReason);
+
+        switch (newState) {
+          case DEGRADED:
+          case SUSPECT:
+            LOGGER.warning(logMessage);
+            break;
+          case HEALTHY:
+            LOGGER.info(logMessage);
```

Comment:
Ideally you don't want to do `info` or `warning` level logs. This can be a `fine` level logging to avoid over-spamming the log files.